react-ui/src/ControlPanel.jsx

```diff
@@ -55,13 +60,28 @@ export class ControlPanel extends PureComponent {
     this.setState({ open: !this.state.open })
   }
 
-  setPictureFilter = ({ value }) => {
+  setPictureFilter = (value) => {
     this.props.setPictureFilter(value)
     this.props.captureUserPicture()
   }
+
+  setPictureFilterWeight = ({ target }) => {
+    this.props.setPictureFilterWeight(target.value)
+    this.props.captureUserPicture()
+  }
+
+  newPictureFilterFromImage = ({ target }) => {
+    const file = target.files[0]
+    imageDataFromFile(file,512)
```

Comment:
So three problems here:

1. There's a bunch of logic nested here: too much for a component to know about.
2. `then`-style asynchronity makes this look sort of bad
3. Inconsistent code style from rest of code (semicolons, missing spaces, etc)

Reply:
To solve this, I'd use redux-thunk to dispatch a different action here, and do the processing there.

```
handleNewFilterImage = (event) => {
  this.props.processNewFilterImage(event.target.files[0])
}

// In actions.js
const processNewFilterImage = (file) => async(dispatch) => {
  const imageData = await imageDataFromFile(file, 512)

  dispatch(setPictureFilter({ label: file.name, value: getImageStyle(imageData) }))
}
```